web/modules/custom/joinup_notification/src/JoinupMessageDelivery.php

```diff
@@ -70,82 +45,146 @@ public function __construct(MessageNotifier $message_notifier) {
+    foreach ($accounts as $account) {
+      // Don't send mails to anonymous users or users that for some reason do
+      // not have an e-mail address set.
+      $mail = $account->getEmail();
+      if ($account->isAnonymous() || empty($mail)) {
+        continue;
+      }
+      // By keying on the user ID we can avoid that a user might get the message
+      // more than once.
+      $recipients_metadata[$account->id()] = [
+        'options' => $notifier_options + ['save on success' => FALSE, 'mail' => $mail],
+        'notifier' => 'email',
+      ];
+    }
+
+    return $this->sendMessage($message, $recipients_metadata);
   }
 
   /**
    * {@inheritdoc}
    */
-  public function createMessage(string $message_template, array $values = []): JoinupMessageDeliveryInterface {
-    // If the template was passed in $values, $message_template take precedence.
-    $values = ['template' => $message_template] + $values;
-    $this->message = Message::create($values);
-    return $this;
+  public function sendMessageToEmailAddresses(MessageInterface $message, array $mails, array $notifier_options = []): bool {
+    $recipients_metadata = [];
+
+    // Ensure uniqueness so that the message is not delivered multiple times to
+    // the same address.
+    foreach (array_unique($mails) as $mail) {
+      $recipients_metadata[] = [
+        'options' => $notifier_options + ['save on success' => FALSE, 'mail' => $mail],
+        'notifier' => 'email',
+      ];
+    }
+
+    return $this->sendMessage($message, $recipients_metadata);
   }
 
   /**
    * {@inheritdoc}
    */
-  public function setArguments(array $arguments): JoinupMessageDeliveryInterface {
-    $this->arguments = $arguments;
-    return $this;
+  public function sendMessageTemplateToMultipleUsers(string $message_template, array $arguments, array $accounts, array $notifier_options = []): bool {
+    $message = $this->createMessage($message_template, $arguments);
+    return $this->sendMessageToMultipleUsers($message, $accounts, $notifier_options);
   }
 
   /**
    * {@inheritdoc}
    */
-  public function setRecipients(array $accounts): JoinupMessageDeliveryInterface {
-    $this->accounts = $accounts;
-    return $this;
+  public function sendMessageTemplateToEmailAddresses(string $message_template, array $arguments, array $mails, array $notifier_options = []): bool {
+    $message = $this->createMessage($message_template, $arguments);
+    return $this->sendMessageToEmailAddresses($message, $mails, $notifier_options);
   }
 
   /**
    * {@inheritdoc}
    */
-  public function setRecipientsAsEmails(array $mails): JoinupMessageDeliveryInterface {
-    $this->mails = $mails;
-    return $this;
+  public function sendMessageTemplateToUser(string $message_template, array $arguments, UserInterface $account, array $notifier_options = [], bool $digest = TRUE): bool {
+    // Don't send messages to the anonymous user.
+    if ($account->isAnonymous()) {
```

Comment:
By not calling this method with an anonymous user account? Are you suggesting we should throw an exception instead?

Reply:
Indeed, Sander meant to throw an exception instead, see above. I didn't realize this when I wrote my previous comment since I read them out of order.

Fixed.